avro-converter/src/main/java/io/confluent/copycat/avro/AvroData.java

```diff
@@ -0,0 +1,1073 @@
+/**
+ * Copyright 2015 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.copycat.avro;
+
+import io.confluent.kafka.serializers.NonRecordContainer;
+
+import org.apache.avro.generic.GenericEnumSymbol;
+import org.apache.avro.generic.GenericFixed;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.GenericRecordBuilder;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.kafka.copycat.data.Field;
+import org.apache.kafka.copycat.data.Schema;
+import org.apache.kafka.copycat.data.SchemaAndValue;
+import org.apache.kafka.copycat.data.SchemaBuilder;
+import org.apache.kafka.copycat.data.Struct;
+import org.apache.kafka.copycat.errors.DataException;
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.node.ArrayNode;
+import org.codehaus.jackson.node.JsonNodeFactory;
+import org.codehaus.jackson.node.ObjectNode;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utilities for converting between our runtime data format and Avro, and (de)serializing that data.
+ */
+public class AvroData {
+  public static final String NAMESPACE = "io.confluent.copycat.avro";
+  // Avro does not permit empty schema names, which might be the ideal default since we also are
+  // not permitted to simply omit the name. Instead, make it very clear where the default is
+  // coming from.
+  public static final String DEFAULT_SCHEMA_NAME = "CopycatDefault";
+  public static final String DEFAULT_SCHEMA_FULL_NAME = NAMESPACE + "." + DEFAULT_SCHEMA_NAME;
+  public static final String MAP_ENTRY_TYPE_NAME = "MapEntry";
+  public static final String KEY_FIELD = "key";
+  public static final String VALUE_FIELD = "value";
+
+  public static final String COPYCAT_NAME_PROP = "copycat_name";
+  public static final String COPYCAT_DOC_PROP = "copycat_doc";
+  public static final String COPYCAT_VERSION_PROP = "copycat_version";
```

Comment:
Huh, I think this was just auto-pilot where I avoided using dots for field names since it's valid but requires you to use json["copycat.name"] style syntax since json.copycat.name would refer to something else. I actually don't like this since I think the `_` is ugly.

There are a couple of options here. We could switch to dots as you say. I could also potentially rework this to put all the extra copycat info into a single `copycat` field that is an object with it's own fields like `name`, `doc`, `version`.